[AIRFLOW-2193] Add ROperator for using R

briandconnelly opened the pull request:
Make sure you have checked _all_ steps below.

### JIRA
- [x] My PR addresses the following [Airflow JIRA](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "[AIRFLOW-XXX] My Airflow PR"
    - https://issues.apache.org/jira/browse/AIRFLOW-2193


### Description
- [x] Here are some details about my PR, including screenshots of any UI changes:

The ROperator allows tasks to be specified using the R programming
language for statistical computing. Tasks can be specified either
as R operations or using a source file. Optionally, the last line of
output can be pushed to an Xcom.

ROperator requires [rpy2](https://rpy2.readthedocs.io/) (and R)

Data can be passed to R using either the environment or through
templating.

### Tests
- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

Several tests are specified in `tests/contrib/operators/test_r_operator.py`, which I cover most of the exposed features

### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

- [x] Passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

Commit:
[AIRFLOW-2193] Add ROperator for using R

The ROperator allows tasks to be specified using the R programming
language for statistical computing. Tasks can be specified either
as R operations or using a source file. Optionally, the last line of
output can be pushed to an Xcom.

ROperator requires that R be installed. By default, it uses the
Rscript interpreter, but littler has also been tested. If Rscript
is not in PATH, the full path can be specified as argument to
`rscript_bin`.

Data can be passed to R using either the environment or through
templating.

Commit:
[AIRFLOW-2193] Add ROperator for using R

The ROperator allows tasks to be specified using the R programming
language for statistical computing. Tasks can be specified either
as R operations or using a source file. Optionally, the last line of
output can be pushed to an Xcom.

ROperator requires that R be installed. By default, it uses the
Rscript interpreter, but littler has also been tested. If Rscript
is not in PATH, the full path can be specified as argument to
`rscript_bin`.

Data can be passed to R using either the environment or through
templating.

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
Update ROperator to use rpy2 instead of CLI

Commit:
Merge branch 'master' of github-personal:briandconnelly/incubator-airflow

Commit:
Merge branch 'master' of github-personal:briandconnelly/incubator-airflow

Commit:
Merge branch 'master' of github-personal:briandconnelly/incubator-airflow

ashb commented on airflow/contrib/operators/r_operator.py:
```diff
@@ -0,0 +1,85 @@
+    def __init__(
+            self,
+            r_command,
+            xcom_push=False,
```
We recently unified this - there is (I think) a `do_xcom_push` flag on BaseOperrator,  so this line, L57 should be removed, and the field  checked on 82 should be `self.do_xcom_push`
```suggestion
```

ashb commented on airflow/contrib/operators/r_operator.py:
```diff
@@ -0,0 +1,85 @@
+
+        super(ROperator, self).__init__(*args, **kwargs)
+        self.r_command = r_command
+        self.xcom_push = xcom_push
```
```suggestion
```

ashb commented on airflow/contrib/operators/r_operator.py:
```diff
@@ -0,0 +1,85 @@
+                    self.log.error("Received R error: %s", e)
+                    res = None
+
+        if self.xcom_push and res:
```
```suggestion
        if self.do_xcom_push and res:
```

ashb commented on tests/contrib/operators/test_r_operator.py:
```diff
@@ -0,0 +1,171 @@
+    def test_xcom_output(self):
+        """Test whether Xcom output is produced using last line"""
+
+        self.task_xcom.xcom_push = True
```
```suggestion
        self.task_xcom.do_xcom_push = True
```

ashb commented on tests/contrib/operators/test_r_operator.py:
```diff
@@ -0,0 +1,171 @@
+    def test_xcom_none(self):
+        """Test whether no Xcom output is produced when push=False"""
+
+        self.task_xcom.xcom_push = False
```
```suggestion
        self.task_xcom.do_xcom_push = False
```

ashb commented on tests/contrib/operators/test_r_operator.py:
```diff
@@ -0,0 +1,171 @@
+        task = ROperator(
+            task_id='test_env_template',
+            r_command='cat(Sys.getenv("{}"))'.format(test_var),
+            env={test_var: "{{ ds }}"},
```
The operator doesn't accept these (yet) right? Looks useful though.

briandconnelly replied:
This is a remnant of a previous version. Originally, the operator ran the R code through the shell via `Rscript`.

ashb committed:
Use do_xcom_push from BaseOperator

Co-Authored-By: briandconnelly <briandconnelly@users.noreply.github.com>

Commit:
Merge branch 'master' of github-personal:briandconnelly/incubator-airflow

Commit:
Remove old ENV var and rscript tests

Commit:
Remove `xcom_push` arg info from docstring

zhongjiajie commented on tests/contrib/operators/test_r_operator.py:
```diff
@@ -0,0 +1,103 @@
+
+from __future__ import print_function, unicode_literals
+
+import os
```
remove `os` package to pass CI test
```suggestion
```

zhongjiajie reviewed (commented):
minor change to pass CI

zhongjiajie committed:
remove `os` package to pass CI test

Co-Authored-By: briandconnelly <briandconnelly@users.noreply.github.com>

zhongjiajie committed:
Add `rpy2` in setup.py to install extras package

briandconnelly committed:
Merge pull request #1 from zhongjiajie/r_support

Add `rpy2` in setup.py to install extras package

briandconnelly committed:
Merge remote-tracking branch 'upstream/master'

briandconnelly committed:
Fix merge issue

dlamblin commented on airflow/contrib/operators/r_operator.py:
```diff
@@ -0,0 +1,79 @@
+                self.log.info("Running command(s):\n%s", self.r_command)
+
+                try:
+                    res = robjects.r.source(fname, echo=False)
```
Just a comment with a question, is it uncommon to write R scripts that take arguments or use environment variables?

The BashOperator has a similar limitation, but can be just flexible enough to take a script written to be called with arguments and have templated arguments used instead with the script as is.

That is you can either write:
```python
bash_command = 'some_command {{params.arg}}',  # this command is templated
bash_command = 'script.sh',  # this file's contents are templated
bash_command = 'script.sh {{params.arg}}', # this file is executable, it is not templated but passed a templated arg.
```

This PR supports the first two cases for R only.

People with pre-existing R scripts, that use arguments, will likely need to rewrite them with templating in mind or fall back to calling their R scripts with BashOperator and templated parameters. Alternatively this operator could have a `r_args` list parameter, but I don't really see a way that [source](https://stat.ethz.ch/R-manual/R-devel/library/base/html/source.html) can accept parameters.
Also [commandArgs](https://stat.ethz.ch/R-manual/R-devel/library/base/html/commandArgs.html) doesn't seem to be something that could be setup for the source.

If you agree that most new and existing users of R in Airflow would end up needing to rewrite their R scripts for this operator, and that's an issue, then this should be addressed (maybe in a later PR, really). And while doing that, it would seem that also taking an `env` dict for setting with [sys.setenv](https://stat.ethz.ch/R-manual/R-devel/library/base/html/Sys.setenv.html) could help in the on-boarding of R tasks into Airflow DAGs..

briandconnelly replied:
I use environment variables quite a bit, so I am planning to re-introduce support for them through an `env` list parameter. They were included in the original version of the operator, but were removed when I switched to rpy2.

Similarly, I'm going to look into passing values to the R environment, which is very similar, but different.

Although I do occasionally use command line arguments for R scripts, since this operator is essentially `source`ing the R code, they seem like they'd be a less natural fit.

briandconnelly replied:
Re-introduced the `env` parameter for exporting environment variables in 
https://github.com/apache/airflow/pull/3115/commits/9f997a20a889fa42a77f0d72184d6dd267d15726

dlamblin replied:
Thanks it looks good.
I had a nit I forgot to mention with the comment referring to pickling xcom. That's a configuration setting, xcom can also be marshaled to json and not pickled. But really… never mind that. LGTM.

briandconnelly replied:
I'm still not feeling 100% about the xcom. I'll update the comment about pickling for sure.

There aren't a lot of operators to use as an example for how to properly respect `do_xcom_push`. The `BashOperator` just unconditionally returns a value, so I followed its lead. But in testing, my tasks always push the value, even when I've set `do_xcom_push=False`.

@ashb, any suggestions?

ashb replied:
That should be okay if you are based off the latest master:

https://github.com/apache/airflow/blob/d46a137ec32eb81a6a2874f62a9623d0bee12c57/airflow/models/__init__.py#L1388-L1392

If _collecting_ the result is expensive then it sometimes makes sense to check `self.do_xcom_push` in the operator (i.e. the docker operator does this as it doesn't make sense to collect the logs and then just throw them away)

shizidushu replied:
@briandconnelly I'm happy to find this ROperator. Thanks. Now I use the DockerOperator to execute R script. But the docker container seems to consumes a lot of memory than I suspect (concurrent tasks).

Just look at the code though I'm not familiar with Python. It seems the code copy R script to a temporary file and then source it. I'm wonder do you consider the following use case: There are three files: `.Rprofile`: specify some setting and functions. `report.Rmd`: generate report file. `send.R`: render the Rmd and send email to someone.

Commit:
Add environment variable support

- Environment variables can be exported using the `env` arg
- Fixed outdated xcom support. Now just returning the result.

mik-laj commented on airflow/contrib/operators/r_operator.py:
```diff
@@ -0,0 +1,99 @@
+
+        # Export context as environment variables
+        airflow_context_vars = context_to_airflow_vars(context, in_env_var_format=True)
+        self.log.info('Exporting the following env vars:\n' +
```
```suggestion
        self.log.info('Exporting the following env vars:\n%s',
```
You should avoid format string before passing to logger.  In some cases, the logger can ignore the text, but the object will be created in memory.

mik-laj committed:
Use logger formatting instead of format string

Co-Authored-By: briandconnelly <briandconnelly@users.noreply.github.com>

ashb commented on airflow/contrib/operators/r_operator.py:
```diff
@@ -0,0 +1,99 @@
+    def __init__(
+            self,
+            r_command,
+            env={},
```
Mutable defaults should be avoided in Python

```suggestion
            env=None,
```

and then in the function

```
    self.env = env or {}
```

ashb reviewed (changes requested):
Could we stub/mock `robjects.r.source` (and maybe more?) so that we don't need an working R environment installed on CI?
